Description: CVE-2022-27943 - limit recursion in rust-demangle
 Add debian/patches/CVE-2022-27943.diff and register it in debian/rules.patch.
 Refreshed for 12.3.0-1ubuntu1~22.04.3 (pr118976 present).
Author: os-patcher
Last-Update: 2026-07-29

--- a/debian/patches/CVE-2022-27943.diff
+++ b/debian/patches/CVE-2022-27943.diff
@@ -0,0 +1,117 @@
+From d1acaa63bc63961d69b337afc9f7a0fe1653a4ce Mon Sep 17 00:00:00 2001
+Date: Mon, 14 Jul 2025 17:09:09 +0530
+Subject: [PATCH] CVE-2022-27943
+
+---
+ src/libiberty/rust-demangle.c | 47 +++++++++++++++++++++++++++++++--------
+ 1 file changed, 38 insertions(+), 9 deletions(-)
+
+diff --git a/src/libiberty/rust-demangle.c b/src/libiberty/rust-demangle.c
+index bb58d900e27..8dfeeac2a4f 100644
+--- a/src/libiberty/rust-demangle.c
++++ b/src/libiberty/rust-demangle.c
+@@ -126,7 +126,7 @@ parse_integer_62 (struct rust_demangler *rdm)
+     return 0;
+ 
+   x = 0;
+-  while (!eat (rdm, '_'))
++  while (!eat (rdm, '_') && !rdm->errored)
+     {
+       c = next (rdm);
+       x *= 62;
+@@ -1082,6 +1082,18 @@ demangle_path_maybe_open_generics (struct rust_demangler *rdm)
+   if (rdm->errored)
+     return open;
+ 
++  if (rdm->recursion != RUST_NO_RECURSION_LIMIT)
++    {
++      ++ rdm->recursion;
++      if (rdm->recursion > RUST_MAX_RECURSION_COUNT)
++	{
++	  /* FIXME: There ought to be a way to report
++	     that the recursion limit has been reached.  */
++	  rdm->errored = 1;
++	  goto end_of_func;
++	}
++    }
++
+   if (eat (rdm, 'B'))
+     {
+       backref = parse_integer_62 (rdm);
+@@ -1107,6 +1119,11 @@ demangle_path_maybe_open_generics (struct rust_demangler *rdm)
+     }
+   else
+     demangle_path (rdm, 0);
++
++ end_of_func:
++  if (rdm->recursion != RUST_NO_RECURSION_LIMIT)
++    -- rdm->recursion;
++
+   return open;
+ }
+ 
+@@ -1148,6 +1165,15 @@ demangle_const (struct rust_demangler *rdm)
+   if (rdm->errored)
+     return;
+ 
++  if (rdm->recursion != RUST_NO_RECURSION_LIMIT)
++    {
++      ++ rdm->recursion;
++      if (rdm->recursion > RUST_MAX_RECURSION_COUNT)
++	/* FIXME: There ought to be a way to report
++	   that the recursion limit has been reached.  */
++	goto fail_return;
++    }
++
+   if (eat (rdm, 'B'))
+     {
+       backref = parse_integer_62 (rdm);
+@@ -1158,7 +1184,7 @@ demangle_const (struct rust_demangler *rdm)
+           demangle_const (rdm);
+           rdm->next = old_next;
+         }
+-      return;
++      goto pass_return;
+     }
+ 
+   ty_tag = next (rdm);
+@@ -1167,7 +1193,7 @@ demangle_const (struct rust_demangler *rdm)
+     /* Placeholder. */
+     case 'p':
+       PRINT ("_");
+-      return;
++      goto pass_return;
+ 
+     /* Unsigned integer types. */
+     case 'h':
+@@ -1200,18 +1226,21 @@ demangle_const (struct rust_demangler *rdm)
+       break;
+ 
+     default:
+-      rdm->errored = 1;
+-      return;
++      goto fail_return;
+     }
+ 
+-  if (rdm->errored)
+-    return;
+-
+-  if (rdm->verbose)
++  if (!rdm->errored && rdm->verbose)
+     {
+       PRINT (": ");
+       PRINT (basic_type (ty_tag));
+     }
++  goto pass_return;
++
++ fail_return:
++  rdm->errored = 1;
++ pass_return:
++  if (rdm->recursion != RUST_NO_RECURSION_LIMIT)
++    -- rdm->recursion;
+ }
+ 
+ static void
+-- 
+2.43.0
+
--- a/debian/rules.patch
+++ b/debian/rules.patch
@@ -70,6 +70,7 @@
 	lp2107313-asan-allocator-base \
 	lp2116909-aarch64-fix-loose-ldpstp-check \
 	pr118976 \
+	CVE-2022-27943 \
 
 ifneq (,$(filter $(distrelease),precise xenial bionic focal groovy hirsute))
   debian_patches += pr100067-revert
